terraso_backend/apps/story_map/permission_rules.py

```diff
@@ -23,12 +27,56 @@ def allowed_to_view_story_map(user, story_map):
 
 @rules.predicate
 def allowed_to_change_story_map(user, story_map):
-    return story_map.created_by == user
+    is_owner = story_map.created_by == user
+    if is_owner:
+        return True
+    account_membership = (
+        story_map.membership_list.memberships.approved_only().filter(user=user).first()
+    )
+    return account_membership is not None and account_membership.user_role == ROLE_COLLABORATOR
 
 
 @rules.predicate
 def allowed_to_delete_story_map(user, story_map):
     return story_map.created_by == user
 
 
+@rules.predicate
+def allowed_to_change_story_map_membership(user, obj):
+    story_map = obj.get("story_map")
+
+    is_owner = story_map.created_by == user
+    if is_owner:
+        return True
+
+    current_membership = obj.get("current_membership")
+    is_new = not current_membership
+    if not is_new:
+        return False
+
+    user_membership = obj.get("user_membership")
+
+    is_collaborator = (
+        user_membership.user_role == ROLE_COLLABORATOR
+        and user_membership.membership_status == Membership.APPROVED
+    )
+    if is_collaborator:
+        return True
+
+    return False
```

Comment:
questions: i'm not sure i understand what is happening in
- `current_membership` is the membership being updated
- `user_membership` is the membership of the user performing the action
- i don't understand why there is a separate check for whether the user is the owner of the story map? can a user who owns a story map be given a different role than collaborator? & if a user who owns a story map has a different role than collaborator, this could says they should still be able to change memberships, but that seems confusing?
- if `current_membership` is the membership being updated, what is the check for `is_new` doing? why would we be doing a permissions check in the first place to mutate a membership that doesn't exist? or if `is_new` doesn't represent whether the membership exists yet, why aren't users allowed to change it if it's new? and why is this check after the story map owner check, so that the owner has different privileges in this regard than other collaborators?

Reply:
@shrouxm I didn't want to get that much in to the details of what ownership means at this time, because that is a bigger conversation, from some conversations with Derek and Hana ownership can be also linked to a terraso group or a landscape in the future, and that opens up a hirarchy in term of what those memberships mean, for example if the ownership of story map moves from a user to a terraso group that means that we will have a membership list linked to the terraso group and a membership list linked to the story map all (potentially) affecting what a user can do with that story map. So it still feels to me that ownership will be a separate piece not linked to membership in the long term, it feels more scalable managing ownership and membership as separate things, potentially it also seems that if the owner of an entity is another entity that has a membership list part of that ownership resolution will be linked to the membership list of the owner.

So that is why I didn't want to get that much into it and at the same time keep ownership separated so that potentially if needed it could evolve that way.

Reply:
:face_with_spiral_eyes: :face_with_spiral_eyes: i see lol that does sound like something that should be resolved separately